File: nodejs/DataBuffer.js
// Creates a resizable Buffer with the given initial length
// length can be omitted, in this case 256 will be used
function DataBuffer(length) {
	this.buffer = new Buffer(length || 256) // allocated buffer
	this.length = 0 // number of used bytes
}

// Appends a byte (uint8 in a number) or a buffer (Buffer or DataBuffer) to this DataBuffer
// Automatically increase the internal buffer size if needed
DataBuffer.prototype.append = function (x) {
	if (typeof x == "number") {
		this._alloc(1)
		this.buffer.writeUInt8(x, this.length)
		this.length++
	} else if (x instanceof Buffer) {
		this._alloc(x.length)
		x.copy(this.buffer, this.length)
		this.length += x.length
	} else if (x instanceof DataBuffer) {
		this._alloc(x.length)
		x.buffer.copy(this.buffer, this.length, 0, x.length)
		this.length += x.length
	} else
		throw new TypeError("number, Buffer or DataBuffer expected")
}

// Makes sure there is enough free space to allocate the given amount of bytes
DataBuffer.prototype._alloc = function (amount) {
	var newBuffer
	if (this.length+amount > this.buffer.length) {
		newBuffer = new Buffer(this.buffer.length*2)
		this.buffer.copy(newBuffer, 0, 0, this.length)
		this.buffer = newBuffer
		this._alloc(amount)
	}
}

module.exports = DataBuffer

File: c#/AsyncProtocol/DataBuffer.cs
using System;

namespace Sitegui.AsyncProtocol {
	/// <summary>
	/// An auto-resizable array for storing binary data
	/// </summary>
	internal class DataBuffer {
		/// <summary>
		/// The internal byte-array for storing all the data
		/// </summary>
		byte[] Buffer;

		/// <summary>
		/// The current number of used bytes in the internal byte-array
		/// </summary>
		int Length = 0;

		/// <summary>
		/// Create a new buffer with the given initial size
		/// </summary>
		/// <param name="length">Number of bytes initially allocated (default: 128)</param>
		public DataBuffer(int length=128) {
			Buffer = new byte[length];
		}

		/// <summary>
		/// Append something to the buffer. Resizes automatically if needed
		/// </summary>
		/// <param name="data">Another DataBuffer instance</param>
		public void Append(DataBuffer data) {
			Alloc(data.Length);
			Array.Copy(data.Buffer, 0, Buffer, Length, data.Length);
			Length += data.Length;
		}

		/// <summary>
		/// Append something to the buffer. Resizes automatically if needed
		/// </summary>
		/// <param name="b">One single byte</param>
		public void Append(byte b) {
			Alloc(1);
			Buffer[Length] = b;
			Length++;
		}

		/// <summary>
		/// Append something to the buffer. Resizes automatically if needed
		/// </summary>
		/// <param name="data">An array of bytes</param>
		public void Append(byte[] data) {
			Alloc(data.Length);
			Array.Copy(data, 0, Buffer, Length, data.Length);
			Length += data.Length;
		}

		/// <summary>
		/// Get this buffer as an array of bytes
		/// </summary>
		/// <returns>Return the byte array</returns>
		public byte[] GetBytes() {
			byte[] r = new byte[Length];
			Array.Copy(Buffer, r, Length);
			return r;
		}

		/// <summary>
		/// Make sure there is at least the given amout of free space. Resize the internal array if needed
		/// </summary>
		/// <param name="amount">The number of free bytes needed</param>
		void Alloc(int amount) {
			while (Length + amount > Buffer.Length) {
				byte[] newBuffer = new byte[Buffer.Length*2];
				Array.Copy(Buffer, newBuffer, Buffer.Length);
				Buffer = newBuffer;
			}
		}
    }
}
